### PR TITLE
Changes end message symbol.

### LINKED ISSUE #7: End message symbol
Communication with the server will be easier when a fixed symbol appears in new line at the end of each message.
For example ```<END>```
Ofcourse clients can also end message with that symbol

### PR DESCRIPTION
**Pull request by devdo-eu**

### CODE REVIEW

**Comment by ppolakiewicz** [comment]
According to documentation default value of encoding parameter for decode and encode methods is 'utf-8' so is it really necessary to be pass everywhere ?

https://docs.python.org/3/library/stdtypes.html#bytes.decode

**Comment by devdo-eu** [comment]
Stack Overflow is to blame ;)
I don't read documentation for `.decode()`. 
Nice tip with this `utf-8`

File: TCP_Server.py [hunk 1]
```diff
@@ -77,11 +77,18 @@ def close_connection(self, conn):
 
     def on_read(self, conn, mask):
         try:
-            data = conn.recv(1000)
+            data = ''
+            limit = 32000
+            while True:
+                limit -= 1
+                data += conn.recv(1).decode('utf-8')
```
Comment (ppolakiewicz) [inline TCP_Server.py]: That looks like pretty long loop.
Can't you read more that one byte in iteration ?

File: dispatcher.py [hunk 2]
```diff
@@ -93,20 +93,20 @@ def run(self):
...
             except OSError as e:
-                logging.info(f"OSError Exception: {e.strerror}")
+                logging.info(f'OSError Exception: {e.strerror}')
         for t in self.threads:
             t.join()
 
     def logBots(self):
         data = self.server.get_data()
         cData = copy.copy(data)
         for k, v in cData.items():
```
Comment (ppolakiewicz) [inline dispatcher.py]: k, v - that is not the best name for variables :)
I know it means key, value but I have no idea what they store.

File: playbot/playbot.py [hunk 3]
```diff
@@ -25,6 +25,8 @@ def __init__(self, host, port):
         self.moves = ['NOP()', 'PATCH()', 'SCAN()', 'OVERLOAD()', 'OVERHEAR()', 'EXPLOIT()', 'INFECT()']
         self.host = host
         self.port = port
+        self.my_move = ''
+        self.move_ok = False
         self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
         self.game = True
```
Comment (ppolakiewicz) [inline playbot/playbot.py]: This is the main variable that determines whether the game works.
I think it is better to init it with False and set to True if connection, login and other nessesery things go right. 
Just look how many times you have to set it to False if something fail.

File: playbot/playbot.py [hunk 4]
```diff
@@ -59,48 +60,55 @@ def name(self):
...
-        except socket.timeout:
-            return b''
+            while True:
+                self.heartbeat_socket()
+                buffor += self.socket.recv(1).decode('utf-8')
+                if '\x04' in buffor:
+                    self.heartbeat = 0
+                    return buffor.split('\x04')[0]
+        except BlockingIOError:
+            self.game = True
```
Comment (ppolakiewicz) [inline playbot/playbot.py:85]: If there is IOException should not it be set to False ?
If I get error on read data then game should not continiue.

File: playbot/playbot.py [hunk 5]
```diff
@@ -59,48 +60,55 @@ def name(self):
...
         except ConnectionResetError as e:
             self.log(e.strerror)
             self.game = False
-            return b''
         except ConnectionAbortedError as e:
             self.log(e.strerror)
             self.game = False
```
Comment (ppolakiewicz) [inline playbot/playbot.py]: Clear buffor when you catch execption so there will not be risk that you will return part of message (for example exception will occure in middle of data read). In play() you check if data startwith which can be true even after you didn't read the whole message.

**Review by ppolakiewicz — COMMENTED** [review]
A few doubts I'd like to discuss.

File: TCP_Server.py (hunk 1 above)
Comment (devdo-eu) [reply TCP_Server.py]: If i read more than one byte - then i will miss `\x04` character and read some more data.
Then next message will be incomplete/invalid.
I can use some kind of side buffer - but i dont like this solution...
If we know how to return read characters to `conn.recv()` buffer - then i can read more.

File: dispatcher.py (hunk 2 above)
Comment (devdo-eu) [reply dispatcher.py]: improved.

File: playbot/playbot.py (hunk 3 above)
Comment (devdo-eu) [reply playbot/playbot.py]: make sense - `self.game` will be initialized as `False`

File: playbot/playbot.py (hunk 4 above)
Comment (devdo-eu) [reply playbot/playbot.py:85]: `IOException` is called when connection is set as `non-blocking` and there are no data to receive from socket.
But it is highly possible that data will be inside socket buffer next time we ask.
Nothing to change here.

File: playbot/playbot.py (hunk 5 above)
Comment (devdo-eu) [reply playbot/playbot.py]: make sense - will be improved.

File: playbot/playbot.py [hunk 6]
```diff
@@ -86,14 +84,15 @@ def get_data(self):
...
-            self.log(e.strerror)
-            self.game = False
+            buffor = e.strerror
         except ConnectionAbortedError as e:
-            self.log(e.strerror)
-            self.game = False
-        return buffor
+            buffor = e.strerror
+
+        self.log(buffor)
```
Comment (ppolakiewicz) [inline playbot/playbot.py]: Please don't use buffor variable to store error message in it.
That's not her purpouse.
Also I'm not a fan of more than one return in method (you have three here) but I can live with that.

If you have identical except block for two or more exceptions then you can combine it to one :)
```
except (ConnectionResetError, ConnectionAbortedError) e:
    self.log(e.strerror)
```
https://docs.python.org/3/tutorial/errors.html#handling-exceptions

Comment (devdo-eu) [reply playbot/playbot.py]: fixed.

**Review by ppolakiewicz — APPROVED** [review]
👍

**Comment by devdo-eu** [comment]
👯  ^^

**Pull request merged**